# [LAB-03] 1. 프로젝트 개요

## #01. 준비작업

### 1. 라이브러리 참조

In [1]:
from jussam import load_data
from helpers import *
from pandas import DataFrame

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/
🔖 Version: 0.5.19


### 2. 데이터 불러오기

In [2]:
origin = load_data("california_housing")
origin.head()

📚 1990년 미국 인구조사(Census)를 기반으로 캘리포니아 각 지역의 인구·주택·소득 특성과 주택 중위가격을 담아, 회귀 분석과 머신러닝 모델 실습에 널리 사용되는 대표적인 주택 가격 데이터 (출처: https://www.kaggle.com/datasets/camnugent/california-housing-prices)

    field               description
--  ------------------  --------------
 0  longitude           경도
 1  latitude            위도
 2  housing_median_age  주택 중위 연령
 3  total_rooms         총 방 수
 4  total_bedrooms      총 침실 수
 5  population          인구
 6  households          가구 수
 7  median_income       중위 소득
 8  median_house_value  주택 중위 가격



,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,-122.230,37.880,41,880,129.000,322,126,8.325,452600,NEAR BAY
1,-122.220,37.860,21,7099,1106.000,2401,1138,8.301,358500,NEAR BAY
2,-122.240,37.850,52,1467,190.000,496,177,7.257,352100,NEAR BAY
3,-122.250,37.850,52,1274,235.000,558,219,5.643,341300,NEAR BAY
4,-122.250,37.850,52,1627,280.000,565,259,3.846,342200,NEAR BAY


## #02. 데이터 구조 점검

### 1. 데이터 크기

In [3]:
print(f"데이터 크기: {origin.shape}")

데이터 크기: (20640, 10)


### 2. 데이터 타입 점검

In [4]:
origin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   longitude           20640 non-null  float64 
 1   latitude            20640 non-null  float64 
 2   housing_median_age  20640 non-null  int64   
 3   total_rooms         20640 non-null  int64   
 4   total_bedrooms      20433 non-null  float64 
 5   population          20640 non-null  int64   
 6   households          20640 non-null  int64   
 7   median_income       20640 non-null  float64 
 8   median_house_value  20640 non-null  int64   
 9   ocean_proximity     20640 non-null  category
dtypes: category(1), float64(4), int64(5)
memory usage: 1.4 MB


### 3. 중복값 점검

In [5]:
df1 = my_qtcheck.check_duplicates(origin)

중복된 행의 수: 0


### 4. 형식·표기·단위 점검 (1) - 연속형 변수

In [6]:
# 값을 검사할 필드 목록
fields = ['housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'median_house_value']

minmax = [] # 최소값과 최대값을 저장할 리스트

for field in fields:
    min_value = df1[field].min()
    max_value = df1[field].max()
    minmax.append({"min": min_value, "max": max_value})

# 최소값과 최대값을 데이터 프레임으로 변환하여 출력
minmax_df = DataFrame(minmax, index=fields)
minmax_df

,min,max
housing_median_age,1.000,52.000
total_rooms,2.000,39320.000
total_bedrooms,1.000,6445.000
population,3.000,35682.000
households,1.000,6082.000
median_income,0.500,15.000
median_house_value,14999.000,500001.000


### 5. 형식·표기·단위 점검 (1) - 명목형 변수

In [7]:
df1['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

## #03. 데이터 품질 점검

### 1. 결측치 확인

In [8]:
my_qtcheck.check_missing_values(df1)

,Missing Count,Missing Ratio (%)
longitude,0,0.000
latitude,0,0.000
housing_median_age,0,0.000
total_rooms,0,0.000
total_bedrooms,207,1.003
population,0,0.000
households,0,0.000
median_income,0,0.000
median_house_value,0,0.000
ocean_proximity,0,0.000


### 2. 결측치 제거

- 결측치 5% 미만이므로 삭제하고 진행

In [9]:
df2 = df1.dropna(subset=["total_bedrooms"])
my_qtcheck.check_missing_values(df2)

,Missing Count,Missing Ratio (%)
longitude,0,0.000
latitude,0,0.000
housing_median_age,0,0.000
total_rooms,0,0.000
total_bedrooms,0,0.000
population,0,0.000
households,0,0.000
median_income,0,0.000
median_house_value,0,0.000
ocean_proximity,0,0.000


### 3. 품질 점검 데이터셋 저장

In [10]:
df2.to_excel("california_housing_qtcheck.xlsx", index=False)

## #04. 기술 통계량

### 1. 범주형 변수의 기술 통계량

In [ ]:
category_desc = my_qtcheck.categorical_summary(df2, 
        save_path="california_housing_categorical_summary.xlsx")
category_desc

📊 컬럼 'ocean_proximity'의 value_counts():


,count
ocean_proximity,
<1H OCEAN,9034
INLAND,6496
ISLAND,5
NEAR BAY,2270
NEAR OCEAN,2628


,ocean_proximity
count,20433
unique,5
top,<1H OCEAN
freq,9034


### 2. 연속형 변수의 기술 통계량

In [14]:
desc_df = my_qtcheck.numerical_summary(df2, 
        save_path="california_housing_numerical_summary.xlsx")
desc_df.T

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,20433.000,20433.000,20433.000,20433.000,20433.000,20433.000,20433.000,20433.000,20433.000
mean,-119.571,35.633,28.633,2636.504,537.871,1424.947,499.433,3.871,206864.413
std,2.004,2.136,12.592,2185.270,421.385,1133.208,382.299,1.899,115435.667
min,-124.350,32.540,1.000,2.000,1.000,3.000,1.000,0.500,14999.000
25%,-121.800,33.930,18.000,1450.000,296.000,787.000,280.000,2.564,119500.000
50%,-118.490,34.260,29.000,2127.000,435.000,1166.000,409.000,3.537,179700.000
75%,-118.010,37.720,37.000,3143.000,647.000,1722.000,604.000,4.744,264700.000
max,-114.310,41.950,52.000,39320.000,6445.000,35682.000,6082.000,15.000,500001.000
rel_diff,-0.009,0.040,0.013,0.240,0.236,0.222,0.221,0.095,0.151
rdiff_flag,similar,similar,similar,diff,diff,diff,diff,similar,diff


#### 💡인사이트

#### (1) 중심 수준 — `rel_diff`로 평균과 중앙값의 거리를 잰다

- `rel_diff = |평균 − 중앙값| ÷ 중앙값` : 평균이 중앙값에서 **상대적으로 얼마나 벗어났는가**.
- 규모 변수(방 수 0.240 · 침실 수 0.236 · 인구 0.222 · 가구 수 0.221)는 평균이 중앙값보다 **22 ~ 24% 크다.**
- 주택 가격 0.151, 중위 소득 0.095, 주택 연령 0.013 순으로 격차가 줄어든다.

#### (2) 퍼짐과 안정성 — `rdiff_flag` 판정에 따라 처리를 나눈다

| 판정 | 기준 | 해당 변수 | 처리 방침 |
|---|---|---|---|
| `similar` | rel_diff < 0.1 | housing_median_age, median_income, longitude, latitude | **평균을 대표값으로 사용 가능** (결측 대체도 평균 허용) |
| `diff` | 0.1 ≤ rel_diff < 0.5 | total_rooms, total_bedrooms, population, households, median_house_value | **대표값·결측 대체 모두 중앙값**, 로그 변환 검토 |
| `large_diff` | rel_diff ≥ 0.5 | 없음 | — |

- `large_diff`가 하나도 없다 → 평균이 완전히 무너진 변수는 없고, **9개 중 5개만 중앙값 기준으로 전환**하면 된다.
- 단 `median_income`은 rel_diff 0.095로 **기준선 0.1 코앞의 경계값**이므로 `similar`로 단정하지 말고 관찰한다.
- 퍼짐의 크기는 변동계수(CV = 표준편차 ÷ 평균)로 확인한다. 규모 변수 **0.77 ~ 0.83(매우 불안정)** > 주택 가격 0.56 > 소득 0.49 > 주택 연령 **0.44(가장 안정)**.
- `diff` + 높은 CV가 겹치는 규모 변수는 **가구당 방 수·가구당 인구** 같은 비율형 파생변수로 정규화한다.

#### (3) 이상치 비율 — 대략적인 예측치

- IQR × 1.5 경계로 계산한 **사전 추정치**다. 실제 제거 전에 규모를 가늠하는 용도로만 본다.
- 상한 이상치: 방 수 6.3% · 침실 수 6.2% · 가구 수 5.9% · 인구 5.8% · 주택 가격 5.2% · 소득 3.3%.
- **하한 이상치는 전 변수 0%**, 위도·경도·주택 연령은 상하한 모두 0%.
- `rdiff_flag = diff`인 변수가 곧 이상치 5% 이상인 변수와 일치한다 → **두 지표가 같은 결론**을 가리킨다.

#### (4) 꼬리 방향

- 위도·경도·주택 연령은 `symmetric`, 나머지는 전부 `right tail` → **우측 꼬리만 존재**한다.
- 규모 변수는 왜도 3.4 ~ 5.0, 첨도 22 ~ 74로 극단적이다. 인구는 중앙값 1,166명 대비 최대 35,682명(**약 31배**) → `log_need = log`.
- `median_income`은 rel_diff는 `similar`이지만 왜도 1.645로 우측 꼬리다 → **중심 지표만으로 분포 형태를 판단할 수 없다.**
- **상한 절단 주의**: 주택 가격 500,001 달러에 985건(4.8%), 주택 연령 52년에 1,265건(6.2%)이 몰려 있다. 분포의 자연스러운 끝이 아니라 조사 시 **잘라낸 값**이므로, 주택 가격 꼬리가 완만해 보이는 것(왜도 0.978)을 그대로 믿으면 안 된다.
- **종합**: 이상치가 전부 상단에만 있으므로 이후 분석은 **상위 이상치 처리(중앙값 기준 전환 · 로그 변환 · 비율형 파생변수 · 절단 인지)** 에 집중한다.